saskatoon/harvest/api.py

```diff
@@ -221,3 +224,281 @@ def list(self, request, *args, **kwargs):
+        }
+
+    def get_total_weight_per_fruit(self, season):
+        treetypes = TreeType.objects.all().order_by("fruit_name")
+        total_weight_harvests_per_fruit = []
+
+        for treetype in treetypes:
+            if season == "all":
+                total_weight = (
+                    HarvestYield.objects.filter(harvest__status="Succeeded")
+                    .filter(tree=treetype)
+                    .aggregate(Sum("total_in_lb"))
+                )
+                total_harvests = (
+                    Harvest.objects.filter(status="Succeeded")
+                    .filter(trees__in=[treetype])
+                    .count()
+                )
+            else:
+                total_weight = (
+                    HarvestYield.objects.filter(harvest__status="Succeeded")
+                    .filter(tree=treetype)
+                    .filter(harvest__start_date__year=season)
+                    .aggregate(Sum("total_in_lb"))
+                )
+                total_harvests = (
+                    Harvest.objects.filter(status="Succeeded")
+                    .filter(trees__in=[treetype])
+                    .filter(start_date__year=season)
+                    .count()
+                )
+
+            if total_weight.get("total_in_lb__sum") is not None:
+                total_weight_harvests_per_fruit.append(
+                    (
+                        treetype.fruit_name,
+                        total_harvests,
+                        total_weight.get("total_in_lb__sum"),
+                    )
+                )
+
+        return total_weight_harvests_per_fruit
+
+    def get_total_weight_per_neighborhood(self, season):
+        neighborhoods = Neighborhood.objects.all().order_by("name")
+        total_weight_harvests_per_neighborhood = []
+
+        for neighborhood in neighborhoods:
+            if season == "all":
+                total_weight = (
+                    HarvestYield.objects.filter(harvest__status="Succeeded")
+                    .filter(harvest__property__neighborhood=neighborhood)
+                    .aggregate(Sum("total_in_lb"))
+                )
+                total_harvests = (
+                    Harvest.objects.filter(status="Succeeded")
+                    .filter(property__neighborhood=neighborhood)
+                    .count()
+                )
+            else:
+                total_weight = (
+                    HarvestYield.objects.filter(harvest__status="Succeeded")
+                    .filter(harvest__property__neighborhood=neighborhood)
+                    .filter(harvest__start_date__year=season)
+                    .aggregate(Sum("total_in_lb"))
+                )
+                total_harvests = (
+                    Harvest.objects.filter(status="Succeeded")
+                    .filter(property__neighborhood=neighborhood)
+                    .filter(start_date__year=season)
+                    .count()
+                )
```

Comment:
@mina-andajos with hindsight I don't think passing the season as parameter is the most efficient, looks like we are filtering the same objects multiple times. Wouldn't it better to fetch the `Harvest` and `HarvestYield` querysets based on season at the top and pass these into the separate functions for further filtering? This way you shouldn't even have to check if season='all' everytime ?

Reply:
thanks for mentioning this i was looking for a way to refactor the functions used in the old saskatoon codebase. that's why I tried to add `self.queryset` before. I'll implement this right away.